lib/logstash/outputs/elasticsearch/data_stream_support.rb

```diff
@@ -0,0 +1,247 @@
+module LogStash module Outputs class ElasticSearch
+  # DS specific behavior/configuration.
+  module DataStreamSupport
+
+    def self.included(base)
+      # Defines whether data will be indexed into an Elasticsearch data stream,
+      # `data_stream_*` settings will only be used if this setting is enabled!
+      # This setting supports values `true`, `false`, and `auto`.
+      # Defaults to `false` in Logstash 7.x and `auto` starting in Logstash 8.0.
+      base.config :data_stream, :validate => ['true', 'false', 'auto']
+
+      base.config :data_stream_type, :validate => ['logs', 'metrics', 'synthetics'], :default => 'logs'
+      base.config :data_stream_dataset, :validate => :dataset_identifier, :default => 'generic'
+      base.config :data_stream_namespace, :validate => :namespace_identifier, :default => 'default'
+
+      base.config :data_stream_sync_fields, :validate => :boolean, :default => true
+      base.config :data_stream_auto_routing, :validate => :boolean, :default => true
+
+      base.extend(Validator)
+    end
+
+    # @override
+    def finish_register
+      super
+
+      if data_stream_config?
+        @event_mapper = -> (e) { data_stream_event_action_tuple(e) }
+        @event_target = -> (e) { data_stream_name(e) }
+        @index = "#{data_stream_type}-#{data_stream_dataset}-#{data_stream_namespace}".freeze # default name
+      end
+    end
+
+    # @note assumes to be running AFTER {after_successful_connection} completed, due ES version checks
+    def data_stream_config?
+      @data_stream_config.nil? ? @data_stream_config = check_data_stream_config! : @data_stream_config
+    end
+
+    private
+
+    def data_stream_name(event)
+      data_stream = event.get('data_stream')
+      return @index if !data_stream_auto_routing || !data_stream.is_a?(Hash)
+
+      type = data_stream['type'] || data_stream_type
+      dataset = data_stream['dataset'] || data_stream_dataset
+      namespace = data_stream['namespace'] || data_stream_namespace
+      "#{type}-#{dataset}-#{namespace}"
+    end
+
+    # @param params the user configuration for the ES output
+    # @note LS initialized configuration (with filled defaults) won't detect as data-stream
+    # compatible, only explicit (`original_params`) config should be tested.
+    # @return [TrueClass|FalseClass] whether given configuration is data-stream compatible
+    def check_data_stream_config!(params = original_params)
+      use_data_stream = data_stream_explicit_value
+      data_stream_params = params.select { |name, _| name.start_with?('data_stream_') } # exclude data_stream =>
+      invalid_data_stream_params = invalid_data_stream_params(params)
+
+      if use_data_stream.eql?(false) && data_stream_params.any?
+        @logger.warn "Ignoring data stream specific configuration (due data_stream => false)", data_stream_params
```

Comment:
I think a refactor dropped my first-pass review bits on this behaviour.

When a user explicitly specifies `data_stream => false`, and also provides `data_stream`-specific settings, I would consider it invalid configuration.

A warning here will easily be missed and recovery of data may be non-trivial, and we can easily prevent the user from accidentally getting themselves in this situation.